File: python-agent/app/api/v1/knowledge.py
import io
from fastapi import APIRouter, UploadFile, File, Form
from pydantic import BaseModel
from app.db.vector_store import VectorStore

router = APIRouter()

vector_store = VectorStore()


class UploadRequest(BaseModel):
    title: str
    content: str
    source_type: str = "text"


@router.post("/knowledge/upload")
async def upload(req: UploadRequest):
    doc_id = await vector_store.add_document(req.title, req.content, req.source_type)
    return {"code": 200, "msg": "success", "data": {"document_id": doc_id}}


def _extract_text(filename: str, content: bytes) -> str:
    ext = filename.rsplit(".", 1)[-1].lower() if "." in filename else ""
    if ext == "docx":
        from docx import Document
        doc = Document(io.BytesIO(content))
        return "\n".join(p.text for p in doc.paragraphs)
    if ext == "pdf":
        import fitz
        text_parts: list[str] = []
        with fitz.open(stream=content, filetype="pdf") as doc:
            for page in doc:
                text_parts.append(page.get_text())
        return "\n".join(text_parts)
    return content.decode("utf-8", errors="replace")


@router.post("/knowledge/upload-file")
async def upload_file(
    title: str = Form(...),
    source_type: str = Form("text"),
    file: UploadFile = File(...),
):
    filename = file.filename or "unknown"
    content = await file.read()
    text = _extract_text(filename, content)
    doc_id = await vector_store.add_document(title, text, source_type)
    return {"code": 200, "msg": "success", "data": {"document_id": doc_id}}


@router.get("/knowledge/documents")
async def list_documents():
    docs = await vector_store.list_sources()
    return {"code": 200, "msg": "success", "data": docs}


@router.delete("/knowledge/documents/{doc_id}")
async def delete_document(doc_id: int):
    ok = await vector_store.delete_document(doc_id)
    if ok:
        return {"code": 200, "msg": "success", "data": None}
    return {"code": 404, "msg": "document not found", "data": None}

File: frontend/src/api/knowledge.ts
const AGENT_URL = '/v1';

export interface KnowledgeDoc {
  source_id: number;
  name: string;
  source_type: string;
  doc_count: number;
  created_at: string;
}

export interface UploadResult {
  document_id: number;
}

export async function uploadDocument(title: string, content: string, sourceType: string = 'text'): Promise<UploadResult> {
  const res = await fetch(`${AGENT_URL}/knowledge/upload`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ title, content, source_type: sourceType }),
  });
  if (!res.ok) throw new Error(`Upload failed: ${res.status}`);
  const json = await res.json();
  return json.data as UploadResult;
}

export async function uploadFile(title: string, file: File, sourceType: string = 'text'): Promise<UploadResult> {
  const form = new FormData();
  form.append('title', title);
  form.append('file', file);
  form.append('source_type', sourceType);
  const res = await fetch(`${AGENT_URL}/knowledge/upload-file`, {
    method: 'POST',
    body: form,
  });
  if (!res.ok) throw new Error(`Upload failed: ${res.status}`);
  const json = await res.json();
  return json.data as UploadResult;
}

export async function listDocuments(): Promise<KnowledgeDoc[]> {
  const res = await fetch(`${AGENT_URL}/knowledge/documents`);
  if (!res.ok) throw new Error(`List failed: ${res.status}`);
  const json = await res.json();
  return json.data as KnowledgeDoc[];
}

export async function deleteDocument(docId: number): Promise<void> {
  const res = await fetch(`${AGENT_URL}/knowledge/documents/${docId}`, { method: 'DELETE' });
  if (!res.ok) throw new Error(`Delete failed: ${res.status}`);
}